cmd/survival/main.go

```diff
@@ -0,0 +1,275 @@
+)
+
+var spellMap = map[spell]string{
+	Undefined:      "",
+	ColdSnap:       "Cold Snap",
+	GhostWalk:      "Ghost Walk",
+	IceWall:        "Ice Wall",
+	Tornado:        "Tornado",
+	EMP:            "EMP",
+	Alacrity:       "Alacrity",
+	SunStrike:      "Sun Strike",
+	ForgeSpirit:    "Forge Spirit",
+	ChaosMeteor:    "Chaos Meteor",
+	DeafeningBlast: "Deafening Blast",
+}
+
+// spellValue
+// map invoker orb spells to invoker skills/spells.
+
+// the reason we use multiplication (*) here instead of addition (+) is that the
+// result wouldn't be unique with addition.
+// So, in this case, we opt for multiplication to ensure a unique value.
+var spellValue = map[rune]spell{
+	'Q' * 'Q' * 'Q': ColdSnap,
+	'Q' * 'Q' * 'W': GhostWalk,
+	'Q' * 'Q' * 'E': IceWall,
+	'W' * 'W' * 'W': EMP,
+	'W' * 'W' * 'Q': Tornado,
+	'W' * 'W' * 'E': Alacrity,
+	'E' * 'E' * 'E': SunStrike,
+	'E' * 'E' * 'Q': ForgeSpirit,
+	'E' * 'E' * 'W': ChaosMeteor,
+	'Q' * 'W' * 'E': DeafeningBlast,
+}
+
+type incantate map[spell]struct{}
+
+// model??? really??? in 2023???
+// think of a better name
+// admittedly this is a copy paste from tutorial
+type model struct {
+	orbs    []string
+	invoked []spell
+	cast    bool
+	// The reason we still keep the `incantate` map is to prevent unnecessary
+	// reordering of the `spells`(slice of string) each time
+	// a casting is triggered (when `d` and `f` are invoked).
+	incantate incantate
+	spells    []string
+	point     int
+	timer     timer.Model
+	isQuit    bool
+}
+
+func main() {
+	m := model{
+		timer:     timer.NewWithInterval(timeout, time.Second),
+		orbs:      make([]string, 3, 3),
+		invoked:   make([]spell, 2, 2),
+		incantate: make(map[spell]struct{}),
+	}
+	m.cast, m.incantate, m.spells = gen(nil)
+	p := tea.NewProgram(m)
+	if _, err := p.Run(); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func (m model) Init() tea.Cmd {
+	// TODO init spells to incantate
+	// can be 2 or 3 spells
+	// can require casting or not
+	return m.timer.Init()
+}
+
+func (m model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
+	switch msg := msg.(type) {
+	case tea.KeyMsg:
+		switch msg.String() {
+		case "ctrl+c":
+			return m, tea.Quit
+		case "q", "w", "e": // handle upper case?
+			m.orbs[2] = m.orbs[1]
+			m.orbs[1] = m.orbs[0]
+			m.orbs[0] = strings.ToUpper(msg.String())
+		case "r":
+			i := invoke(m.orbs)
+			if i == Undefined {
+				break
+			}
+			// if stage only requires invoking, we can still count?
+			if !m.cast {
+				if _, ok := m.incantate[i]; ok {
+					delete(m.incantate, i)
+					m.point++
+
+					m.spells = reOrder(m.spells, spellMap[i])
+				}
+			}
+			if m.invoked[0] == i {
+				break
+			}
+			m.invoked[1] = m.invoked[0]
+			m.invoked[0] = i
+		case "d":
+			// Invoke here to prevent unnecessary reordering of spells every time a user triggers
+			// a spell casting.
+			if _, ok := m.incantate[m.invoked[0]]; ok {
+				delete(m.incantate, m.invoked[0])
+				m.point++
+				m.timer.Timeout += 2 * time.Second
```

Comment:
not privy to how it actually works in invokergame, but seems like each set of spells have assigned timeout, instead of adding to existing time. looks like the logic belongs in `gen`? ditto for the treatment in `case "f"`below

Reply:
same as above, i made the timeout reset to 10 seconds when spell invoked or cast, but i don't think that same as in invokergame